test: organize melos for full local testing, minor fixes so complete runs quickly pass

mikehardy opened the pull request:
## Description

As a new contributor, it is difficult to say that you have a correct working environment before making your first change, or that after your change you have not broken anything.

The goal of this PR is purely local test oriented then
- add highly parallel targets to melos.yaml that allow for a fast local runs of the various quality checks
- add one high level melos.yaml target that can be trusted to fail if anything is wrong
- fix all problems encountered while exercising those tasks

## Related Issues

This is related to the forward-port PR #4249 but these commits may be considered independently and immediately to help developers now, and to shrink that PR to it's core purpose

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

Salakar commented on packages/firebase_messaging/firebase_messaging/example/test_driver/instance_e2e.dart:
```diff
@@ -89,16 +89,17 @@ void runInstanceTests() {
-        expect(token1, isNot(token2));
-      }, skip: SKIP_MANUAL_TESTS); // only run for manual testing
-    });
+    // FIXME `melos run tests:e2e` with ios simulator: [E]: MissingPluginException(No implementation found for method Messaging#deleteToken on channel plugins.flutter.io/firebase_messaging)
```
Oh my gosh, I see the problem here, quite a bad one actually, will cause issues with `getInitialMessage`, `deleteToken`, `getAPNSToken` & `setForegroundNotificationPresentationOptions`;

This line should be an `else if` https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/ios/Classes/FLTFirebaseMessagingPlugin.m#L117-L118 but it's not so it's running the method specific code AND also falling back to the final `else` on the next `if else` blocks which causes the not implemented: https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/ios/Classes/FLTFirebaseMessagingPlugin.m#L143

Sending up a PR for this shortly

Salakar replied:
PR #4339 merged, so you should no longer need this

mikehardy replied:
Fantastic! As I respond your comments I'll edit this out before re-pushing

mikehardy replied:
dropped this commit before re-pushing

Salakar commented on packages/firebase_ml_vision/example/test_driver/firebase_ml_vision_e2e.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:e2e/e2e.dart';
```
Can leave these two files for ml vision out once global ignore added to melos.yaml

mikehardy replied:
I wasn't able to get it in global ignore but I was able to ignore it just for one e2e target, as the others didn't match selection criteria

Salakar reviewed (changes requested):
Minor changes

mikehardy commented on melos.yaml:
```diff
@@ -4,6 +4,10 @@ packages:
   - packages/**
 
```
:bulb:

mikehardy replied:
@Salakar I cannot get a global ignore to work in this way at all

I add:

```
ignore:
  - "*firebase_ml_custom*"
```

...at the top of the file and it has zero affect on `melos run build:example_ios` whereas it works on the target. Just for grins I tried ignores of `*`, `**`, `*firebase*` etc just to see if the global ignore would have any influence at all and it did not seem to ever change the total package count. It would change `melos list` output though. :thinking: :shrug: 

Is there something interacting with application of global ignore such that it wouldn't work?

After I repush this (in a few minutes after grooming what I can) you might pull this branch and try it.

mikehardy replied:
after carefully checking selection criteria, it seems I only need to have this in a couple targets, since global ignore wasn't working, I put it in where needed and it works locally again.

mikehardy commented on packages/firebase_admob/example/test_driver/firebase_admob_e2e.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:e2e/e2e.dart';
```
will look after we figure out global ignore not working...

mikehardy replied:
I wasn't able to get it in global ignore but I was able to ignore it just for one e2e target, as the others didn't match selection criteria

mikehardy commented on packages/firebase_admob/example/test_driver/firebase_admob_e2e_test.dart:
```diff
@@ -0,0 +1,3 @@
+import 'package:e2e/e2e_driver.dart' as e2e;
```
I wasn't able to get it in global ignore but I was able to ignore it just for one e2e target, as the others didn't match selection criteria

mikehardy commented:
@Salakar down to 4 commits now, with a cleaner ignores list. Suggest handling global ignores issue as a separate item

mikehardy commented:
Okay, I rebased this to upstream/master and it seemed like for safety's sake I should include the commit that fixes firebase_ml_vision so it works with the final 25.x.x Android BoM (note that commit is battle-tested from react-native-firebase - you may check related links on it - it is ugly but correct)

I also fixed the issue where from a clean checkout the `melos run qualitycheck` works around the `pub get`/`--no-pub` issue where things aren't initialized fully after `melos bootstrap` so the whole thing really does work all the way through.

Unless auth rejects you for too many requests but that's out of my control :shrug: 

I'd love to see this or something like this integrated, it is the test harness I use to prove the firebase-ios-sdk v7+ / firebase-android-sdk v26+ #4249  work and auth emulator work #4263 

Cheers :-)

mikehardy commented:
Until all the E2E workflow components that rely on network resources are wrapped in re-tries, the e2e jobs here will be subject to flakiness, it's not the PR that caused the firebase_messaging android failure, but I can't re-trigger that

```

[firebase_messaging_example]: Running Gradle task 'assembleDebug'...                          
[firebase_messaging_example]:   108.1s
[firebase_messaging_example]: 
[firebase_messaging_example]: 	at org.gradle.wrapper.Install.createDist(Install.java:48)
[firebase_messaging_example]: 	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:128)
[firebase_messaging_example]: 	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
[firebase_messaging_example]: 	Suppressed: java.net.SocketException: Broken pipe (Write failed)
[firebase_messaging_example]: 		at java.net.SocketOutputStream.socketWrite0(Native Method)
[firebase_messaging_example]: 		at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
[firebase_messaging_example]: 		at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
[firebase_messaging_example]: 		at sun.security.ssl.SSLSocketOutputRecord.encodeAlert(SSLSocketOutputRecord.java:81)
[firebase_messaging_example]: 		at sun.security.ssl.TransportContext.fatal(TransportContext.java:355)
[firebase_messaging_example]: 		at sun.security.ssl.TransportContext.fatal(TransportContext.java:267)
[firebase_messaging_example]: 		at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:397)
[firebase_messaging_example]: 		... 13 more
[firebase_messaging_example]: Caused by: java.io.EOFException: SSL peer shut down incorrectly
[firebase_messaging_example]: 	at sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:167)
[firebase_messaging_example]: 	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:108)
[firebase_messaging_example]: 	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1143)
[firebase_messaging_example]: 	... 15 more
[firebase_messaging_example]: Gradle task assembleDebug failed with exit code 1
```

Salakar reviewed (approved):
LGTM!

![LGTM GIF](https://i.kym-cdn.com/photos/images/newsfeed/001/287/658/263.gif)
